src/field_types.c

```diff
@@ -253,86 +136,54 @@ field_types_init_descriptors(int num_field_types, field_type *ft)
-        if (tup == NULL) {
-            goto finish;
-        }
-        int res = PyList_Append(dtype_list, tup);
-        Py_DECREF(tup);
-        if (res < 0) {
-            goto finish;
-        }
-    }
-    PyArray_DescrConverter(dtype_list, &result);
-
-  finish:
-    Py_XDECREF(empty_string);
-    Py_XDECREF(dtype_list);
-    return result;
-}
-
-#ifdef TESTMAIN
-
 void
-show_field_types(int num_fields, field_type *ft)
-{
-    field_types_fprintf(stdout, num_fields, ft);
-
-    bool homogeneous = field_types_is_homogeneous(num_fields, ft);
-    printf("homogeneous = %s\n", homogeneous ? "true" : "false");
-
-    int32_t total_size = field_types_total_size(num_fields, ft);
-    printf("total_size = %lu\n", (unsigned long)total_size);
-
-    char *dtypestr = field_types_build_str(num_fields, NULL, homogeneous, ft);
-    printf("dtypestr = '%s'\n", dtypestr);
-
-    free(dtypestr);
+field_types_clear(int num_field_types, field_type *ft) {
+    for (int i = 0; i < num_field_types; i++) {
+        Py_XDECREF(ft[i].descr);
+        ft[i].descr = NULL;
+    }
 }
 
 
-int
-main(int argc, char *argv[])
+field_type *
+field_types_create(int num_field_types, PyArray_Descr *dtypes[])
 {
-    char *codes = "ffHHSU";
-    int32_t sizes[] = {8, 8, 2, 2, 4, 48};
-    int num_fields = sizeof(sizes) / sizeof(sizes[0]);
-
-    field_type *ft = field_types_create(num_fields, codes, sizes);
-    printf("ft:\n");
-    show_field_types(num_fields, ft);
-
-    int status = field_types_grow(num_fields + 2, num_fields, &ft);
-    ft[num_fields].typecode = 'b';
-    ft[num_fields].itemsize = 1;
-    ++num_fields;
-    ft[num_fields].typecode = 'f';
-    ft[num_fields].itemsize = 4;
-    ++num_fields;
-
-    printf("\n");
-    printf("ft, after grow:\n");
-    show_field_types(num_fields, ft);
-
-    free(ft);
+    field_type *ft;
 
-    return 0;
+    ft = malloc(num_field_types * sizeof(field_type));
+    if (ft == NULL) {
+        return NULL;
+    }
+    for (int i = 0; i < num_field_types; ++i) {
+        PyArray_Descr *descr = dtypes[i];
+        if (PyDataType_ISSIGNED(descr)) {
+            ft[i].typecode = 'i';
+        }
+        else if (PyDataType_ISUNSIGNED(descr)) {
+            ft[i].typecode = 'u';
+        }
+        else if (PyDataType_ISFLOAT(descr) && descr->elsize <= 8) {
+            ft[i].typecode = 'f';
+        }
+        else if (PyDataType_ISCOMPLEX(descr) && descr->elsize <= 16) {
+            ft[i].typecode = 'c';
+        }
+        else if (descr->type_num == NPY_STRING) {
+            ft[i].typecode = 'S';
+        }
+        else if (descr->type_num == NPY_UNICODE) {
+            ft[i].typecode = 'U';
+        }
+        else {
+            ft[i].typecode = 'x';
+        }
+        ft[i].itemsize = descr->elsize;
+        Py_INCREF(descr);
+        ft[i].descr = descr;
+    }
+    if (field_types_init_descriptors(num_field_types, ft) < 0) {
+        free(ft);
+        return NULL;
+    }
+    return ft;
```

Comment:
The code changes (as opposed to deletions), are simply rearranging the functions because the now first function can be `static` (and does not need to be included in the header file).